### PR TITLE
Allow vanitisation

### PR DESCRIPTION
**Pull request by TheoLeanse**
@matthew-andrews 

Does this make sense?

### CODE REVIEW

File: client/lib/authorised-fetch.js [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+	const fetchOpts = {
+		method: 'POST',
+		headers: {
+			'FT-Authorization': auth,
+			'Content-Type': 'application/json'
+		},
+		body: JSON.stringify(reqBody)
+	};
+	return fetch(url, fetchOpts);
+}
```
Comment (matthew-andrews) [inline client/lib/authorised-fetch.js]: nice idea to centralise — especially as you could have some event emitted if you get a unauthorised response at this point to deauthorise the user.

File: client/components/create.js [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+		return {};
+	},
+	putVanity(code) {
+		return () => {
+			const fromURL = document.querySelector("#fromURL").value;
+			const toURL = document.querySelector("#toURL").value;
+			if (!fromURL || !toURL) {
+				return this.setState({message: 'Please enter \'from\' and \'to\' fields.'});
+			}
+			authorisedFetch({
```
Comment (matthew-andrews) [inline client/components/create.js]: 🇺🇸 english in code plz 😁

File: client/components/create.js [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+				fromURL,
+				toURL,
+				code
+			}).then(res => {
+				if (res.ok) {
+					return this.setState({message: 'That went OK!'});
+				}
+				throw 'Bad response';
+			}).catch(() => {
+				this.setState({message: 'Not OK!'});
```
Comment (matthew-andrews) [inline client/components/create.js]: perhaps a `// TODO` to handle the errors

**Comment by matthew-andrews** [comment]
Yes makes very much sense - some very minor suggestions.

I think I need to build you a test environment tomorrow.

File: client/components/create.js [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+		return (
+			<div className="o-forms-group">
+				<h1 className="ums-section-header">
+					Create a vanity URL or a redirect
+				</h1>
+				<label className="o-forms-label" for="fromURL">From:</label>
+				<input className="o-forms-text" type="text" id="fromURL"/>
+				<label className="o-forms-label" for="toURL">To:</label>
+				<input className="o-forms-text" type="text" id="toURL"/>
+				<button className="o-buttons" onClick={ this.putVanity(100) }>
```
Comment (matthew-andrews) [inline client/components/create.js]: possibly might make sense for this function to be called `putUrl`

**Comment by TheoLeanse** [comment]
Thanks for those suggestions!

File: client/components/check.js [hunk 5]
```diff
@@ -33,12 +22,16 @@ const Check = React.createClass({
 			.catch(err => console.log(err));
```
Comment (matthew-andrews) [inline client/components/check.js]: perhaps a todo to handle the error^^

Comment (matthew-andrews) [inline client/components/check.js]: (e.g. if you put in `https://www.ft.com/stream/brandId/NTlhNzEyMzMtZjBjZi00Y2U1LTg0ODUtZWVjNmEyYmU1NzQ2-QnJhbmRz` into the field and click *Check* you don't get any feedback)

File: client/lib/authorised-fetch.js (hunk 1 above)
Comment (TheoLeanse) [inline client/lib/authorised-fetch.js]: Do you mean something like...
```js
return fetch(url, fetchOpts)
        .then(res => {
                if (res.status === 401) {
                        \\ remove cookies and redirect page
                }
                return res;
        };
```
or to emit an event in this function which gets handled elsewhere?

Comment (matthew-andrews) [inline client/lib/authorised-fetch.js]: i think emitting an event is a slightly cleaner way of doing this.

this code can just handle it as if it were a normal error and then dispatch an event on the body, that a global handler can handle.

File: client/lib/authorized-fetch.js [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+			'Content-Type': 'application/json'
+		},
+		body: JSON.stringify(reqBody)
+	};
+	return fetch(url, fetchOpts)
+		.catch(error => {
+			if (error.message === 'Failed to fetch') {
+				document.dispatchEvent(new Event('authError'));
+			}
+			return error;
```
Comment (TheoLeanse) [inline client/lib/authorized-fetch.js]: @matthew-andrews 
so, a bit like this?
(is it necessary to pass errors along or maybe  throw them to allow callers to handle them?)

Comment (matthew-andrews) [inline client/lib/authorized-fetch.js]: Yep, hm, it's a bit annoying that it's not possible to be more fine grained than this.  I might make a support request to Amazon to ask for API Gateway authentication errors to have (configurable) CORS headers.

But this is 👌 for now.

Comment (matthew-andrews) [inline client/lib/authorized-fetch.js]: ah yes you'll need to rethrow that error or `return Promise.reject(error)`

File: client/lib/authorized-fetch.js [hunk 7]
```diff
@@ -0,0 +1,21 @@
+import reactCookie from 'react-cookie';
+
+export default function authorisedFetch (reqBody) {
+	const auth = reactCookie.load('s3o_username') + ' ' + reactCookie.load('s3o_token');
+	const url = "https://vadkwcngik.execute-api.eu-west-1.amazonaws.com/prod/";
```
Comment (matthew-andrews) [inline client/lib/authorized-fetch.js]: should be update this to the test env `n89xyc146c.execute-api.eu-west-1.amazonaws.com`?

**Pull request merged**